source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,630 @@
+                     ( const char * ) pxAzureIoTHubClient->_internal.pucHostname ) );
+        xResult = eAzureIoTHubClientSuccess;
+    }
+
+    return xResult;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                           const uint8_t * pucTelemetryData,
+                                                           uint32_t ulTelemetryDataLength,
+                                                           AzureIoTMessageProperties_t * pxProperties )
+{
+    AzureIoTMQTTResult_t xMQTTResult;
+    AzureIoTHubClientResult_t xResult;
+    AzureIoTMQTTPublishInfo_t xMQTTPublishInfo = { 0 };
+    uint16_t usPublishPacketIdentifier;
+    size_t xTelemetryTopicLength;
+    az_result xCoreResult;
+
+    if( pxAzureIoTHubClient == NULL )
+    {
+        AZLogError( ( "Failed to send telemetry: Invalid argument." ) );
+        xResult = eAzureIoTHubClientInvalidArgument;
+    }
+    else if( az_result_failed( xCoreResult = az_iot_hub_client_telemetry_get_publish_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                                            ( pxProperties != NULL ) ? &pxProperties->_internal.xProperties : NULL,
+                                                                                            ( char * ) pxAzureIoTHubClient->_internal.pucWorkingBuffer,
+                                                                                            pxAzureIoTHubClient->_internal.ulWorkingBufferLength,
+                                                                                            &xTelemetryTopicLength ) ) )
+    {
+        AZLogError( ( "Failed to get telemetry topic: 0x%08x.", xCoreResult ) );
+        xResult = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        xMQTTPublishInfo.xQOS = eAzureIoTMQTTQoS1;
+        xMQTTPublishInfo.pcTopicName = pxAzureIoTHubClient->_internal.pucWorkingBuffer;
+        xMQTTPublishInfo.usTopicNameLength = ( uint16_t ) xTelemetryTopicLength;
+        xMQTTPublishInfo.pvPayload = ( const void * ) pucTelemetryData;
+        xMQTTPublishInfo.xPayloadLength = ulTelemetryDataLength;
+
+        /* Get a unique packet id. */
+        usPublishPacketIdentifier = AzureIoTMQTT_GetPacketId( &( pxAzureIoTHubClient->_internal.xMQTTContext ) );
+
+        /* Send PUBLISH packet. */
+        if( ( xMQTTResult = AzureIoTMQTT_Publish( &( pxAzureIoTHubClient->_internal.xMQTTContext ),
+                                                  &xMQTTPublishInfo, usPublishPacketIdentifier ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "Failed to publish telementry: 0x%08x.", xMQTTResult ) );
+            xResult = eAzureIoTHubClientPublishFailed;
+        }
+        else
+        {
+            xResult = eAzureIoTHubClientSuccess;
```

Comment:
Nit: Other places after going to AzureIoTMQTT_* layer on success, there's a LogInfo().  I think one should be here too.